components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/service/IdentityKeyStoreGeneratorImpl.java

```diff
@@ -0,0 +1,334 @@
+     * <p>
+     * This method creates a new KeyStore for the specified tenant domain and context if it does not already exist.
+     * It ensures thread safety by using a lock mechanism to avoid concurrent modifications when creating the KeyStore.
+     * If the KeyStore for the given context already exists, the method exits without performing any operations.
+     * </p>
+     *
+     * @param tenantDomain the tenant domain for which the KeyStore is to be created.
+     * @param context      the context for which the KeyStore is to be generated.
+     * @throws KeyStoreManagementException if an error occurs during KeyStore creation or initialization.
+     */
+    public void generateContextKeyStore(String tenantDomain, String context) throws KeyStoreManagementException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        this.keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        this.tenantDomain = tenantDomain;
+        ReentrantLock lock = tenantLocks.computeIfAbsent(tenantId, id -> new ReentrantLock());
+        boolean lockAcquired = false; // Track if the lock was acquired
+
+        try {
+            IdentityTenantUtil.initializeRegistry(tenantId);
+            if (isContextKeyStoreExists(context)) {
+                return; // KeyStore already exists, no need to create again
+            }
+
+            lock.lock(); // Acquire the lock
+            lockAcquired = true; // Mark the lock as acquired
+
+            if (!isContextKeyStoreExists(context)) {
+                // Create the KeyStore
+                password = generatePassword();
+                KeyStore keyStore = KeystoreUtils.getKeystoreInstance(
+                        KeystoreUtils.getKeyStoreFileType(tenantDomain));
+                keyStore.load(null, password.toCharArray());
+                generateContextKeyPair(keyStore, context);
+                persistContextKeyStore(keyStore, context);
+            }
+
+        } catch (Exception e) {
+            String msg = "Error while instantiating a keystore";
+            LOG.error(msg, e);
+            throw new KeyStoreManagementException(msg, e);
+        } finally {
+            if (lockAcquired) { // Only release the lock if it was acquired
+                lock.unlock();
+            }
+            tenantLocks.remove(tenantId); // Clean up locks for this tenant if not needed anymore
+        }
+    }
+
+    private boolean isContextKeyStoreExists(String context) {
+
+        String keyStoreName = KeystoreUtils.getKeyStoreFileLocation(tenantDomain +
+                KEY_STORE_CONTEXT_SEPARATOR + context);
+        boolean isKeyStoreExists = false;
+        try {
+            keyStoreManager.getKeyStore(keyStoreName);
+            isKeyStoreExists = true;
+        } catch (Exception e) {
+            String msg = "Error while checking the existence of keystore.  ";
+            LOG.error(msg + e.getMessage());
```

Comment:
Do we need to log the error message here? I think this existence check is made before creating the keystore. So we can make this a debug log. Additionally we are catching the Exception class so is it guaranteed that only the exception that we are expecting is caught here. If not make the catch blog specific.

Reply:
fixed with https://github.com/wso2/carbon-identity-framework/pull/6189/commits/0477800848f16b5fc2a59cdd4696f788d7a4b225